File: internal/infrastructure/adapters/mongodb/adapter.go
// Package mongodb provides MongoDB adapter for the storage library
package mongodb

import (
	"context"
	"fmt"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/AnandSGit/HybridCache.io/internal/domain"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
	"go.mongodb.org/mongo-driver/mongo/readpref"
)

// Adapter implements the domain.Adapter interface for MongoDB
type Adapter struct {
	name    string
	version string
}

// NewAdapter creates a new MongoDB adapter
func NewAdapter() domain.Adapter {
	return &Adapter{
		name:    "mongodb",
		version: "1.0.0",
	}
}

// Name returns the adapter name
func (a *Adapter) Name() string {
	return a.name
}

// Version returns the adapter version
func (a *Adapter) Version() string {
	return a.version
}

// DatabaseType returns the database type
func (a *Adapter) DatabaseType() domain.DatabaseType {
	return domain.DatabaseTypeMongoDB
}

// Connect creates a new storage instance
func (a *Adapter) Connect(ctx context.Context, config domain.Config) (domain.Storage, error) {
	if err := a.ValidateConfig(config); err != nil {
		return nil, domain.NewConnectionError("INVALID_CONFIG", err.Error()).WithCause(err)
	}

	uri, err := a.buildURI(config)
	if err != nil {
		return nil, domain.NewConnectionError("INVALID_URI", "failed to build MongoDB URI").WithCause(err)
	}

	clientOptions := options.Client().ApplyURI(uri)

	// Configure connection pool
	clientOptions.SetMaxPoolSize(uint64(config.MaxOpenConns))
	clientOptions.SetMinPoolSize(uint64(config.MaxIdleConns))
	clientOptions.SetMaxConnIdleTime(config.ConnMaxIdleTime)
	clientOptions.SetConnectTimeout(config.ConnectTimeout)
	clientOptions.SetSocketTimeout(config.QueryTimeout)

	// Create MongoDB client
	client, err := mongo.Connect(ctx, clientOptions)
	if err != nil {
		return nil, domain.NewConnectionError("CONNECTION_FAILED", "failed to connect to MongoDB").WithCause(err)
	}

	// Test the connection
	if err := client.Ping(ctx, readpref.Primary()); err != nil {
		client.Disconnect(ctx)
		return nil, domain.NewConnectionError("PING_FAILED", "failed to ping MongoDB").WithCause(err)
	}

	storage := &MongoDBStorage{
		client:   client,
		database: client.Database(config.Database),
		adapter:  a,
		config:   config,
	}

	return storage, nil
}

// ParseDSN parses a MongoDB connection string
func (a *Adapter) ParseDSN(dsn string) (domain.Config, error) {
	config := domain.Config{
		DSN: dsn,
	}

	// Parse MongoDB URI
	u, err := url.Parse(dsn)
	if err != nil {
		return config, fmt.Errorf("invalid MongoDB URI: %w", err)
	}

	// Extract basic connection info
	if u.Hostname() != "" {
		config.Host = u.Hostname()
	}
	if u.Port() != "" {
		if port, err := strconv.Atoi(u.Port()); err == nil {
			config.Port = port
		}
	}
	if u.User != nil {
		config.Username = u.User.Username()
		if password, ok := u.User.Password(); ok {
			config.Password = password
		}
	}

	// Extract database name from path
	if len(u.Path) > 1 {
		config.Database = strings.TrimPrefix(u.Path, "/")
	}

	// Parse query parameters
	params := u.Query()
	if len(params) > 0 {
		if config.Options == nil {
			config.Options = make(map[string]interface{})
		}
		for key, values := range params {
			if len(values) > 0 {
				config.Options[key] = values[0]
			}
		}
	}

	// Set defaults
	if config.Port == 0 {
		config.Port = 27017
	}
	if config.MaxOpenConns == 0 {
		config.MaxOpenConns = 100
	}
	if config.MaxIdleConns == 0 {
		config.MaxIdleConns = 10
	}
	if config.ConnMaxLifetime == 0 {
		config.ConnMaxLifetime = 1 * time.Hour
	}
	if config.ConnMaxIdleTime == 0 {
		config.ConnMaxIdleTime = 30 * time.Minute
	}
	if config.ConnectTimeout == 0 {
		config.ConnectTimeout = 10 * time.Second
	}
	if config.QueryTimeout == 0 {
		config.QueryTimeout = 30 * time.Second
	}

	return config, nil
}

// ValidateConfig validates the configuration
func (a *Adapter) ValidateConfig(config domain.Config) error {
	if config.DSN == "" {
		if config.Host == "" {
			return fmt.Errorf("host is required")
		}
		if config.Database == "" {
			return fmt.Errorf("database is required")
		}
	}

	if config.MaxOpenConns <= 0 {
		return fmt.Errorf("max open connections must be positive")
	}

	if config.MaxIdleConns < 0 {
		return fmt.Errorf("max idle connections cannot be negative")
	}

	if config.MaxIdleConns > config.MaxOpenConns {
		return fmt.Errorf("max idle connections cannot exceed max open connections")
	}

	return nil
}

// buildURI builds a MongoDB connection URI from config
func (a *Adapter) buildURI(config domain.Config) (string, error) {
	if config.DSN != "" {
		return config.DSN, nil
	}

	// Build URI from individual components
	var uri strings.Builder
	uri.WriteString("mongodb://")

	// Add authentication if provided
	if config.Username != "" {
		uri.WriteString(url.QueryEscape(config.Username))
		if config.Password != "" {
			uri.WriteString(":")
			uri.WriteString(url.QueryEscape(config.Password))
		}
		uri.WriteString("@")
	}

	// Add host and port
	uri.WriteString(config.Host)
	if config.Port != 0 && config.Port != 27017 {
		uri.WriteString(":")
		uri.WriteString(strconv.Itoa(config.Port))
	}

	// Add database
	if config.Database != "" {
		uri.WriteString("/")
		uri.WriteString(config.Database)
	}

	// Add options
	if len(config.Options) > 0 {
		uri.WriteString("?")
		var params []string
		for key, value := range config.Options {
			params = append(params, fmt.Sprintf("%s=%s", key, url.QueryEscape(fmt.Sprintf("%v", value))))
		}
		uri.WriteString(strings.Join(params, "&"))
	}

	return uri.String(), nil
}

// TranslateQuery translates a domain.Query to MongoDB operations
func (a *Adapter) TranslateQuery(query domain.Query) (string, []interface{}, error) {
	// MongoDB doesn't use SQL, so we'll return the query as-is for now
	// In a full implementation, this would translate SQL-like queries to MongoDB operations
	return query.SQL, query.Parameters, nil
}

// TranslateCommand translates a domain.Command to MongoDB operations
func (a *Adapter) TranslateCommand(command domain.Command) (string, []interface{}, error) {
	// MongoDB doesn't use SQL, so we'll return the command as-is for now
	// In a full implementation, this would translate SQL-like commands to MongoDB operations
	return command.SQL, command.Parameters, nil
}

// MapGoType maps Go types to MongoDB data types
func (a *Adapter) MapGoType(goType interface{}) (domain.DataType, error) {
	switch goType.(type) {
	case string:
		return domain.DataTypeString, nil
	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
		return domain.DataTypeInteger, nil
	case float32, float64:
		return domain.DataTypeFloat, nil
	case bool:
		return domain.DataTypeBoolean, nil
	case time.Time:
		return domain.DataTypeDateTime, nil
	case []byte:
		return domain.DataTypeBinary, nil
	case map[string]interface{}, []interface{}:
		return domain.DataTypeJSON, nil
	default:
		return domain.DataTypeUnknown, fmt.Errorf("unsupported Go type: %T", goType)
	}
}

// MapDatabaseType maps MongoDB BSON types to domain data types
func (a *Adapter) MapDatabaseType(dbType string) (domain.DataType, error) {
	switch strings.ToLower(dbType) {
	case "string":
		return domain.DataTypeString, nil
	case "int", "long", "int32", "int64":
		return domain.DataTypeInteger, nil
	case "double", "decimal":
		return domain.DataTypeFloat, nil
	case "bool", "boolean":
		return domain.DataTypeBoolean, nil
	case "date", "timestamp":
		return domain.DataTypeDateTime, nil
	case "bindata", "binary":
		return domain.DataTypeBinary, nil
	case "object", "document":
		return domain.DataTypeJSON, nil
	case "array":
		return domain.DataTypeArray, nil
	case "objectid":
		return domain.DataTypeString, nil // ObjectID as string
	default:
		return domain.DataTypeUnknown, fmt.Errorf("unsupported MongoDB type: %s", dbType)
	}
}

// SupportsTransactions returns whether MongoDB supports transactions
func (a *Adapter) SupportsTransactions() bool {
	return true // MongoDB 4.0+ supports multi-document transactions
}

// SupportsJoins returns whether MongoDB supports joins
func (a *Adapter) SupportsJoins() bool {
	return true // MongoDB supports $lookup aggregation for joins
}

// SupportsBatch returns whether MongoDB supports batch operations
func (a *Adapter) SupportsBatch() bool {
	return true
}

// SupportsSchema returns whether MongoDB supports schema operations
func (a *Adapter) SupportsSchema() bool {
	return true // MongoDB supports collections and indexes
}

File: python/src/storage/adapters/mongodb/adapter.py
"""
MongoDB adapter implementation.

This module provides the main adapter class for MongoDB database connections
using Motor for high-performance async operations.
"""

from __future__ import annotations

import re
from typing import Any, List
from urllib.parse import parse_qs, urlparse

from storage.adapters.mongodb.connection import MongoDBStorage
from storage.errors import ValidationError, new_validation_error
from storage.interfaces import Adapter, Storage
from storage.types import (
    Command,
    Config,
    DataType,
    DatabaseType,
    Query,
)


class MongoDBAdapter(Adapter):
    """MongoDB database adapter using Motor."""
    
    def name(self) -> str:
        """Get adapter name."""
        return "mongodb"
    
    def version(self) -> str:
        """Get adapter version."""
        return "1.0.0"
    
    def database_type(self) -> str:
        """Get database type."""
        return DatabaseType.MONGODB.value
    
    async def connect(self, config: Config) -> Storage:
        """Create a storage connection.
        
        Args:
            config: Connection configuration
            
        Returns:
            Storage instance
            
        Raises:
            ConnectionError: If connection fails
        """
        self.validate_config(config)
        storage = MongoDBStorage(config)
        await storage.connect()
        return storage
    
    def parse_dsn(self, dsn: str) -> Config:
        """Parse a MongoDB connection string.
        
        Args:
            dsn: MongoDB connection string
            
        Returns:
            Parsed configuration
            
        Raises:
            ValidationError: If DSN is invalid
        """
        try:
            parsed = urlparse(dsn)
            
            config = Config(
                dsn=dsn,
                host=parsed.hostname or "localhost",
                port=parsed.port or 27017,
                database=parsed.path.lstrip("/") if parsed.path else "",
                username=parsed.username or "",
                password=parsed.password or "",
            )
            
            # Parse query parameters
            if parsed.query:
                params = parse_qs(parsed.query)
                options = {}
                
                # Common MongoDB connection options
                for key, values in params.items():
                    if values:
                        value = values[0]
                        if key in ["authSource", "replicaSet", "ssl", "tls"]:
                            options[key] = value
                        elif key in ["maxPoolSize", "minPoolSize", "maxIdleTimeMS", "connectTimeoutMS", "socketTimeoutMS"]:
                            try:
                                options[key] = int(value)
                            except ValueError:
                                pass
                
                if options:
                    config.options = options
            
            return config
            
        except Exception as e:
            raise new_validation_error(f"Invalid MongoDB DSN: {e}") from e
    
    def validate_config(self, config: Config) -> None:
        """Validate the configuration.
        
        Args:
            config: Configuration to validate
            
        Raises:
            ValidationError: If configuration is invalid
        """
        if not config.dsn:
            if not config.host:
                raise new_validation_error("host is required")
            if not config.database:
                raise new_validation_error("database is required")
        
        if config.max_open_conns <= 0:
            raise new_validation_error("max_open_conns must be positive")
        
        if config.max_idle_conns < 0:
            raise new_validation_error("max_idle_conns cannot be negative")
        
        if config.max_idle_conns > config.max_open_conns:
            raise new_validation_error("max_idle_conns cannot exceed max_open_conns")
    
    def translate_query(self, query: Query) -> tuple[str, List[Any]]:
        """Translate a query to MongoDB-specific operations.
        
        Args:
            query: Query to translate
            
        Returns:
            Tuple of (operation, parameters)
            
        Raises:
            QueryError: If translation fails
        """
        # MongoDB doesn't use SQL, so we return the query as-is for now
        # In a full implementation, this would translate SQL-like queries to MongoDB operations
        return query.sql, query.parameters
    
    def translate_command(self, command: Command) -> tuple[str, List[Any]]:
        """Translate a command to MongoDB-specific operations.
        
        Args:
            command: Command to translate
            
        Returns:
            Tuple of (operation, parameters)
            
        Raises:
            QueryError: If translation fails
        """
        # MongoDB doesn't use SQL, so we return the command as-is for now
        # In a full implementation, this would translate SQL-like commands to MongoDB operations
        return command.sql, command.parameters
    
    def map_go_type(self, go_type: str) -> DataType:
        """Map Go types to MongoDB data types.
        
        Args:
            go_type: Go type name
            
        Returns:
            Corresponding data type
            
        Raises:
            ValidationError: If type mapping fails
        """
        type_mapping = {
            "string": DataType.STRING,
            "int": DataType.INTEGER,
            "int8": DataType.INTEGER,
            "int16": DataType.INTEGER,
            "int32": DataType.INTEGER,
            "int64": DataType.INTEGER,
            "uint": DataType.INTEGER,
            "uint8": DataType.INTEGER,
            "uint16": DataType.INTEGER,
            "uint32": DataType.INTEGER,
            "uint64": DataType.INTEGER,
            "float32": DataType.FLOAT,
            "float64": DataType.FLOAT,
            "bool": DataType.BOOLEAN,
            "time.Time": DataType.DATETIME,
            "[]byte": DataType.BINARY,
            "map[string]interface{}": DataType.JSON,
            "[]interface{}": DataType.ARRAY,
        }
        
        if go_type in type_mapping:
            return type_mapping[go_type]
        
        raise new_validation_error(f"Unsupported Go type: {go_type}")
    
    def map_database_type(self, db_type: str) -> DataType:
        """Map MongoDB BSON types to domain data types.
        
        Args:
            db_type: MongoDB BSON type name
            
        Returns:
            Corresponding data type
            
        Raises:
            ValidationError: If type mapping fails
        """
        type_mapping = {
            "string": DataType.STRING,
            "int": DataType.INTEGER,
            "long": DataType.INTEGER,
            "int32": DataType.INTEGER,
            "int64": DataType.INTEGER,
            "double": DataType.FLOAT,
            "decimal": DataType.DECIMAL,
            "bool": DataType.BOOLEAN,
            "boolean": DataType.BOOLEAN,
            "date": DataType.DATETIME,
            "timestamp": DataType.DATETIME,
            "bindata": DataType.BINARY,
            "binary": DataType.BINARY,
            "object": DataType.JSON,
            "document": DataType.JSON,
            "array": DataType.ARRAY,
            "objectid": DataType.STRING,  # ObjectID as string
        }
        
        normalized_type = db_type.lower()
        if normalized_type in type_mapping:
            return type_mapping[normalized_type]
        
        raise new_validation_error(f"Unsupported MongoDB type: {db_type}")
    
    def supports_transactions(self) -> bool:
        """Check if MongoDB supports transactions."""
        return True  # MongoDB 4.0+ supports multi-document transactions
    
    def supports_joins(self) -> bool:
        """Check if MongoDB supports joins."""
        return True  # MongoDB supports $lookup aggregation for joins
    
    def supports_batch(self) -> bool:
        """Check if MongoDB supports batch operations."""
        return True
    
    def supports_schema(self) -> bool:
        """Check if MongoDB supports schema operations."""
        return True  # MongoDB supports collections and indexes
    
    def _build_connection_uri(self, config: Config) -> str:
        """Build MongoDB connection URI from config.
        
        Args:
            config: Database configuration
            
        Returns:
            MongoDB connection URI
        """
        if config.dsn:
            return config.dsn
        
        # Build URI from components
        uri_parts = ["mongodb://"]
        
        # Add authentication
        if config.username:
            uri_parts.append(config.username)
            if config.password:
                uri_parts.append(f":{config.password}")
            uri_parts.append("@")
        
        # Add host and port
        uri_parts.append(config.host)
        if config.port and config.port != 27017:
            uri_parts.append(f":{config.port}")
        
        # Add database
        if config.database:
            uri_parts.append(f"/{config.database}")
        
        # Add options
        if config.options:
            query_params = []
            for key, value in config.options.items():
                query_params.append(f"{key}={value}")
            if query_params:
                uri_parts.append(f"?{'&'.join(query_params)}")
        
        return "".join(uri_parts)
